File: src/m_concurrency/concurrency.java
import m_concurrency.WeatherData;
import java.util.ArrayList;
import java.util.List;
import static m_concurrency.ConcurrencyKt.parseWeatherData;

void main() {
    List<String> cities = List.of("New York", "London", "Paris", "Rome", "Madrid", "Vienna", "Moscow");

    // Thread to print every second
    Thread timerThread = new Thread(() -> {
        try {
            for (int i = 1; i <= 10; i++) {
                Thread.sleep(1000);
                System.out.println(STR."\{i} seconds passed");
            }
        } catch (InterruptedException e) {
            e.printStackTrace();
        }
    });
    timerThread.start();

    updateWeatherUI(cities);

    try {
        timerThread.join();
    } catch (InterruptedException e) {
        e.printStackTrace();
    }
}

public static class WeatherFetcher implements Runnable {
    private final String location;
    private final WeatherDataListener listener;

    public WeatherFetcher(String location, WeatherDataListener listener) {
        this.location = location;
        this.listener = listener;
    }

    @Override
    public void run() {
        try {
            Thread.sleep(5000);
        } catch (InterruptedException e) {
            throw new RuntimeException(e);
        }
        String data = STR."Weather data for \{location}";
        WeatherData weatherData = parseWeatherData(data);
        listener.onDataFetched(weatherData);
    }
}

public interface WeatherDataListener {
    void onDataFetched(WeatherData data);
}

public void updateWeatherUI(List<String> locations) {
    List<Thread> threads = new ArrayList<>();
    List<WeatherData> weatherDataList = new ArrayList<>();

    for (String location : locations) {
        WeatherFetcher fetcher = new WeatherFetcher(location, data -> {
            synchronized (weatherDataList) {
                weatherDataList.add(data);
                if (weatherDataList.size() == locations.size()) {
                    System.out.println();
                    System.out.println(STR."weatherDataList = \{weatherDataList}");
                    System.out.println();
                }
            }
        });
        Thread thread = new Thread(fetcher);
        threads.add(thread);
        thread.start();
    }

    for (Thread thread : threads) {
        try {
            thread.join();
        } catch (InterruptedException e) {
            e.printStackTrace();
        }
    }
}

File: src/m_concurrency/concurrency.kt
package m_concurrency
//"org.jetbrains.kotlinx:kotlinx-coroutines-core-jvm:1.9.0"

import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Deferred
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.async
import kotlinx.coroutines.awaitAll
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch
import kotlinx.coroutines.runBlocking
import kotlinx.coroutines.withContext


suspend fun foo() {
    fetchWeatherData("Berlin")
}

fun main(args: Array<String>) = runBlocking {
    val cities =
        listOf("Berlin", "Tokyo", "New York", "London", "Paris", "Rome", "Madrid", "Vienna", "Moscow", "Beijing")


    fetchWeatherData("Berlin")

    val job = CoroutineScope(Dispatchers.Default).launch {

        //val deferredData1 = CoroutineScope(Dispatchers.Default).async { fetchWeatherData("Berlin") }
        //println(deferredData1.await())

        val deferredData = cities.map { async { fetchWeatherData(it) } }

        val weatherData = deferredData.awaitAll()
        println()
        println(weatherData)
        println()
    }

    for (i in 1..10) {
        delay(1000)
        println("$i seconds passed")
    }

    job.join()
}

suspend fun fetchWeatherData(location: String): WeatherData {
    val data = withContext(Dispatchers.IO) {
        // make a network request to fetch weather data for the given location
        delay(5000)
        "Weather data for $location"
    }
    return parseWeatherData(data)
}

fun parseWeatherData(data: String): WeatherData {
    return WeatherData(data)
}